### PR TITLE
Rt images

### LINKED ISSUE #42: [feature] Render target saving to and loading from image
**Describe the solution you'd like**
`IRenderTarget:Load(filename)` which would load the contents of the file into the render target, resizing the RT as needed, and `IRenderTarget:Save(filename, mip = 0)` which would save the contents of the render target at the specified mip to a file.  
The file path would be relative to `garrysmod/data/vistrace` and should sanitise out things like `/../` (should be easy to do with the filesystem library).  

**Describe alternatives you've considered**
Instead of scaling the RT to fit an image, it may be better to scale the image to fit the RT and add a`vistrace.LoadImage()` function to automatically size the RT.  
This would allow loading images at a consistent resolution for post processing, or loading an image into a specific MIP level.

### PR DESCRIPTION
**Pull request by yogwoggf**
Fixes #42.

**PR Type (tick all that are applicable)**
- [ ] Bug Fix
- [x] New Feature (doesn't break Module <-> Lua compatibility)
- [ ] New Feature (breaks Module <-> Lua compatibility)
- [x] Documentation Update Required

**Tested Targets (only applicable for changes to the binary module, delete otherwise)**
- [x] windows-x64-release
- [ ] windows-x86-release (currently not possible)

**Checklist**
- [x] I have read and understand the contributing guidelines
- [x] I have tested all aspects of this PR (not required)
- [ ] I have added the watermark on the branding branch with my desired name to any content attached where applicable (only applies to showcase submissions)

**Description**
When merged, this PR will allow users to save images into real files (hdr and png) and load images into rendertargets.

**Features**
- [x] Saving and loading
- [x] No escape of file path sandboxing (tested with most methods)
- [ ] vistrace.LoadImage (will be added soon with stb_image_resize)

### CODE REVIEW

File: include/vistrace/IRenderTarget.h [hunk 1]
```diff
@@ -62,6 +62,9 @@ namespace VisTrace
 		virtual Pixel GetPixel(uint16_t x, uint16_t y, uint8_t mip = 0) const = 0;
 		virtual void SetPixel(uint16_t x, uint16_t y, const Pixel& pixel, uint8_t mip = 0) = 0;
 
+		virtual bool Load(const char* filename) = 0;
+		virtual bool Write(const char* filename, uint8_t mips = 0) = 0;
```
Comment (Derpius) [inline include/vistrace/IRenderTarget.h:66]: Should be `mip`, not `mips`, also should be `Save` not `Write`

File: source/VisTrace.cpp [hunk 2]
```diff
@@ -228,6 +229,53 @@ LUA_FUNCTION(RT_GenerateMIPs)
...
+{
+	LUA->CheckType(1, RenderTarget::id);
+	LUA->CheckType(2, Type::String);
+
+	RenderTarget* pRt = *LUA->GetUserType<RenderTarget*>(1, RenderTarget::id);
+
+	if (!pRt->IsValid()) LUA->ThrowError("Invalid render target");
+
+	const char* filename = LUA->GetString(2);
+	if (!filename) {
```
Comment (Derpius) [inline source/VisTrace.cpp:242]: You don't need this check, if the type was a string and this returned a nullptr, the engine's Lua API itself is broken

File: source/VisTrace.cpp [hunk 3]
```diff
@@ -228,6 +229,53 @@ LUA_FUNCTION(RT_GenerateMIPs)
...
+		LUA->ThrowError("Invalid file name");
+	}
+
+	bool loaded = pRt->Load(filename);
+	if (!loaded) LUA->ThrowError("Failed to load image into render target, invalid image file?");
+
+	return 0;
+}
+
+LUA_FUNCTION(RT_Write)
```
Comment (Derpius) [inline source/VisTrace.cpp:253]: Should be `Save`

File: source/VisTrace.cpp [hunk 4]
```diff
@@ -228,6 +229,53 @@ LUA_FUNCTION(RT_GenerateMIPs)
...
+{
+	LUA->CheckType(1, RenderTarget::id);
+	LUA->CheckType(2, Type::String);
+
+	uint8_t mips = 0;
```
Comment (Derpius) [inline source/VisTrace.cpp:258]: Should be `mip`

File: source/VisTrace.cpp [hunk 5]
```diff
@@ -228,6 +229,53 @@ LUA_FUNCTION(RT_GenerateMIPs)
...
+	if (LUA->IsType(3, Type::Number)) { // Optional mips argument
```
Comment (Derpius) [inline source/VisTrace.cpp:259]: Don't need a manual typecheck, `GetNumber` returns 0 (the default) on failure

File: source/VisTrace.cpp [hunk 6]
```diff
@@ -228,6 +229,53 @@ LUA_FUNCTION(RT_GenerateMIPs)
...
+		mips = LUA->GetNumber(3);
+	}
+
+	RenderTarget* pRt = *LUA->GetUserType<RenderTarget*>(1, RenderTarget::id);
+	if (!pRt->IsValid()) LUA->ThrowError("Invalid render target");
+
+	const char* filename = LUA->GetString(2);
+	if (!filename) {
```
Comment (Derpius) [inline source/VisTrace.cpp:267]: Don't need to check this, see above

File: source/VisTrace.cpp [hunk 7]
```diff
@@ -1011,6 +1059,11 @@ GMOD_MODULE_OPEN()
 		LUA->PushCFunction(RT_SetPixel);
 		LUA->SetField(-2, "SetPixel");
 
+		LUA->PushCFunction(RT_Load);
+		LUA->SetField(-2, "Load");
+		LUA->PushCFunction(RT_Write);
+		LUA->SetField(-2, "Write");
```
Comment (Derpius) [inline source/VisTrace.cpp:1065]: Should be `Save`

File: source/objects/RenderTarget.cpp [hunk 8]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+		}
+
+		bool resized = Resize(width, height);
+		if (!resized) {
+			stbi_image_free(imageBuffer);
+			return false;
+		}
+
+		memcpy(GetRawData(), imageBuffer, stride * width * height * channels);
+		GenerateMIPs();
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:240]: Useless call, you're resizing with 1 mip on line 233

File: source/objects/RenderTarget.cpp [hunk 9]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+		stbi_image_free(imageBuffer);
+
+		return true;
+	}
+	else if (stride == sizeof(float)) {
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:245]: You don't need to duplicate *all* of the above code, just change the load call as needed and use `uint8_t` for the image buffer either way

File: source/objects/RenderTarget.cpp [hunk 10]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+		GenerateMIPs();
+		stbi_image_free(imageBuffer);
+
+		return true;
+	}
+
+	return false; // wtf? invalid stride?
+}
+
+bool RenderTarget::Write(const char* filename, uint8_t mips) {
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:268]: Should be `Save`, and `mip`

File: source/objects/RenderTarget.cpp [hunk 11]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+	uint8_t* imageData = GetRawData(mips);
+
+	size_t stride = STRIDES[static_cast<uint8_t>(GetFormat())];
+	uint8_t channels = CHANNELS[static_cast<uint8_t>(GetFormat())];
+
+	const bool isFloat = stride == sizeof(float);
+
+	const char* extension = isFloat ? ".hdr" : ".png";
+
+	fs::path workingDir = fs::current_path();
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:278]: Are you sure this is gonna be correct at all times?
There are easy ways to get the executable and/or this dll's location...

File: source/objects/RenderTarget.cpp [hunk 12]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+	workingDir += "/garrysmod/data/";
+	workingDir = workingDir.make_preferred(); // Convert to forwardslashes on Windows and backslashes on POSIX.
+
+	fs::path filenamePath(filename);
+	filenamePath = filenamePath.make_preferred();
+
+	// Do not allow going up the file tree
+	const char* transformedFilename = filenamePath.string().c_str();
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:286]: You should just get the abs path with `filesystem` and check the root vs the `data/vistrace` folder

File: source/objects/RenderTarget.cpp [hunk 13]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+	if (filenamePath.has_parent_path()) {
+		fs::path tempPath(workingDir);
+		tempPath += filenamePath;
+		tempPath.lexically_normal();
+		tempPath = tempPath.parent_path();
+
+		fs::create_directories(tempPath);
+	}
+
+	workingDir += filenamePath;
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:306]: You're re-adding the filename path onto the root over and over again, you should just be appending the possibly relative path to the working dir, converting to absolute, checking the root is unchanged, and finally creating subdirs/saving

File: source/objects/RenderTarget.cpp [hunk 14]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+	workingDir = workingDir.lexically_normal();
+	workingDir.replace_extension(extension);
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:308]: The extension passed by the user should **never** be replaced, either the filetype should be inferred from the extension and matched to the RT format, or if that fails, **append** an appropriate extension to the existing file path
This allows for saving a file like `frame.001` without discarding the `.001`

File: source/objects/RenderTarget.cpp [hunk 15]
```diff
@@ -209,6 +214,119 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
...
+
+	const char* newPath = workingDir.string().c_str();
+
+	// Surely nothing can go wrong
```
Comment (Derpius) [inline source/objects/RenderTarget.cpp:312]: Bit of a nitpick, but you have comments like this all over your PR
We're not Valve

File: source/objects/RenderTarget.h [hunk 16]
```diff
@@ -57,5 +57,8 @@ class RenderTarget : public VisTrace::IRenderTarget
 	VisTrace::Pixel GetPixel(uint16_t x, uint16_t y, uint8_t mip = 0) const;
 	void SetPixel(uint16_t x, uint16_t y, const VisTrace::Pixel& pixel, uint8_t mip = 0);
 
+	bool Load(const char* filename);
+	bool Write(const char* filename, uint8_t mips = 0);
```
Comment (Derpius) [inline source/objects/RenderTarget.h:61]: Should be `Save`

**Review by Derpius — CHANGES_REQUESTED** [review]

File: source/objects/RenderTarget.cpp (hunk 15 above)
Comment (yogwoggf) [reply source/objects/RenderTarget.cpp:312]: Its never too big to dream

**Pull request merged**